features/nvstore/source/nvstore.cpp

```diff
@@ -0,0 +1,1066 @@
+int32_t safe_increment(uint32_t &value, uint32_t increment)
+{
+    return core_util_atomic_incr_u32(&value, increment);
+}
+
+// Check whether a buffer is aligned.
+// Parameters :
+// buf           - [IN]   Data buffer.
+// size          - [IN]   Alignment size.
+// Return        : Boolean result.
+static inline int is_buf_aligned(const void *buf, uint32_t size)
+{
+    return (((size_t) buf / size * size) == (size_t) buf);
+}
+
+// Align a value to a specified size.
+// Parameters :
+// val           - [IN]   Value.
+// size          - [IN]   Size.
+// Return        : Aligned value.
+static inline uint32_t align_up(uint32_t val, uint32_t size)
+{
+    return (((val-1) / size) + 1) * size;
+}
+
+// CRC32 calculation. Supports "rolling" calculation (using the initial value).
+// Parameters :
+// init_crc      - [IN]   Initial CRC.
+// data_len      - [IN]   Buffer's data length.
+// data_buf      - [IN]   Data buffer.
+// Return        : CRC.
+static uint32_t crc32(uint32_t init_crc, uint32_t data_len, uint8_t *data_buf)
+{
+    uint32_t i, j;
+    uint32_t crc, mask;
+
+    crc = init_crc;
+    for (i = 0; i < data_len; i++) {
+       crc = crc ^ (uint32_t) (data_buf[i]);
+       for (j = 0; j < 8; j++) {
+          mask = -(crc & 1);
+          crc = (crc >> 1) ^ (0xEDB88320 & mask);
+       }
+    }
+    return crc;
+}
+
+NVStore::NVStore() : _init_done(0), _init_attempts(0), _active_area(0), _max_keys(NVSTORE_MAX_KEYS),
+               _active_area_version(0), _free_space_offset(0), _size(0), _offset_by_key(0)
+{
+}
+
+NVStore::~NVStore()
+{
+    if (_init_done) {
+        deinit();
+    }
+}
+
+uint16_t NVStore::get_max_keys() const
+{
+    return _max_keys;
+}
+
+void NVStore::set_max_keys(uint16_t num_keys)
+{
+    MBED_ASSERT(num_keys < MASTER_RECORD_KEY);
+    _max_keys = num_keys;
+    // User is allowed to change number of types. As this affects init, need to deinitialize now.
+    // Don't call init right away - it is lazily called by get/set functions if needed.
+    deinit();
+}
+
+// Flash access helper functions, using area and offset notations
+
+// Read from flash, given area and offset.
+// Parameters :
+// area          - [IN]   Flash area.
+// offset        - [IN]   Offset in area.
+// len_bytes     - [IN]   Length in bytes.
+// buf           - [IN]   Data buffer.
+// Return        : 0 on success. Error code otherwise.
+int NVStore::flash_read_area(uint8_t area, uint32_t offset, uint32_t len_bytes, uint32_t *buf)
+{
+    return nvstore_int_flash_read(len_bytes, _flash_area_params[area].address + offset, buf);
+}
+
+// Write to flash, given area and offset.
+// Parameters :
+// area          - [IN]   Flash area.
+// offset        - [IN]   Offset in area.
+// len_bytes     - [IN]   Length in bytes.
+// buf           - [IN]   Data buffer.
+// Return        : 0 on success. Error code otherwise.
+int NVStore::flash_write_area(uint8_t area, uint32_t offset, uint32_t len_bytes, const uint32_t *buf)
```

Comment:
Even if this is private, the documentation for API should be in a header file, not a code file, can you move it there and do it in the style you did other public API?

Reply:
Will do.

Reply:
0xc0170  - not sure if we want full doxy style comments for these functions. It will then be added to our documentation, which is redundant IMO. I believe that a short explanation (as @sg- suggested) is more appropriate here.

Reply:
I believe doxygen does not show private members/methods by default. I find documenting them useful for developer (thus only stays in the implementation, not exposed to documentation).